props/saber_BC_buttons.h

```diff
@@ -386,31 +1134,317 @@ EFFECT(volmax);     // for maximum volume reached
+    mode::getSL<SPEC>()->SayBlade();
+    mode::getSL<SPEC>()->SayWhole(current_blade_);
+  }
+
+  void select() override {
+    // Set the current blade for the ChangeBladeLengthMode before transitioning
+    // Deactivate current mode and push to LED length adjustment mode for the selected blade
+    selected_blade_ = current_blade_;
+    highlighted_blade_.Stop(current_blade_);
+    SPEC::SteppedMode::mode_deactivate();
+    pushMode<typename SPEC::ChangeBladeLengthMode>();
+  }
+
+  void exit() override {
+    highlighted_blade_.Stop(current_blade_);
+    SPEC::MenuBase::exit();
+  }
+};
+#endif  // BC_DUAL_BLADES
+
+
+template<class SPEC>
+struct BCChangeBladeLengthMode : public SPEC::MenuBase {
+  virtual int blade() { return selected_blade_; }
+
+  int steps_per_revolution() override {
+    return 40;  // adjust for sensitivity
+  }
+
+  void mode_activate(bool onreturn) override {
+    int len = prop_GetBladeLength(blade());
+    int maxlen = prop_GetMaxBladeLength(blade());
+    saved_len_ = len;
+    if (len == -1) len = maxlen;
+    this->pos_ = len;
+    prop_SetBladeLength(blade(), maxlen);
+    showlen_.Start(blade());
+    SPEC::SteppedMode::mode_activate(onreturn);
+    // mode::getSL<SPEC>()->SayEditBladeLength();
+    say();
+  }
+
+  void mode_deactivate() {
+    showlen_.Stop(blade());
+  }
+
+  void say() override {
+    mode::getSL<SPEC>()->SayWhole(this->pos_);
+  }
+
+  uint16_t size() override {
+    return prop_GetMaxBladeLength(blade());
+  }
+
+  void select() override {
+    prop_SetBladeLength(blade(), this->pos_);
+    prop_SaveState();
+    mode::getSL<SPEC>()->SaySave();
+    SPEC::SteppedMode::select();
+  }
+
+  void exit() override {
+    prop_SetBladeLength(blade(), saved_len_);
+    SPEC::MenuBase::exit();
+  }
+
+  void update() override {
+    hybrid_font.PlayPolyphonic(&SFX_volup);
+    if (!SFX_volup) SaberBase::sound_length = 0.2;
+    this->say_time_ = Cyclint<uint32_t>(millis()) + (uint32_t)(SaberBase::sound_length * 1000) + 400;
+    if (!this->say_time_) this->say_time_ += 1;
+    this->fadeout(SaberBase::sound_length);
+  }
+
+  int getLength() { return this->pos_; }
+
+  ShowColorSingleBladeTemplate<typename SPEC::ShowLengthStyle> showlen_;
+  int saved_len_;
+};
+
+#endif  // DYNAMIC_BLADE_LENGTH
+
+template<class SPEC>
+struct BCMenuSpec {
+  typedef BCVolumeMode<SPEC> BCVolumeMenu;
+#ifdef DYNAMIC_BLADE_LENGTH
+  typedef BCSelectBladeMode<SPEC> SelectBladeMode;
+  typedef mode::ShowLengthStyle<SPEC> ShowLengthStyle;
+  typedef BCChangeBladeLengthMode<SPEC> ChangeBladeLengthMode;
+#endif
+  typedef mode::SelectCancelMode SelectCancelMode;
+  typedef mode::SteppedMode<SPEC> SteppedMode;
+  typedef mode::MenuBase<SPEC> MenuBase;
+  typedef SoundLibraryV2Template<SPEC> SoundLibrary;
+};
+#endif  // BC_ENABLE_OS_MENU
+
+#include "prop_base.h"
+
+#ifdef MENU_SPEC_TEMPLATE
+#undef MENU_SPEC_TEMPLATE
+#endif
+
+#ifdef BC_ENABLE_OS_MENU
+#define MENU_SPEC_TEMPLATE DefaultMenuSpec
```

Comment:
MENU_SPEC_TEMPLATE is meant to be controlled by the config file.
How about just doing:

#ifndef MENU_SPEC_TEMPLATE
#define MENU_SPEC_TEMPLATE BCMenuSpec
#endif


?

Reply:
ok let me try to explain.
This was trying to workaround guarded code in prop_base. Maybe that code isn't expecting custom specs, but I needed sound library polled (now init() ) but it is inside `ifdef MENU_SPEC_TEMPLATE`.
I thought to just satisfy that by defining MENU_SPEC_TEMPLATE to BCMenuSpec at the beginning of my prop file, but then it complains that TopMenu isn't in the custom spec due to how EnterMenu() gets set.

So the workaround was to first do `#define MENU_SPEC_TEMPLATE DefaultMenuSpec` at the top of my prop to satisfy the prop_base stuff, then undef it, then re-define it according to BC_ENABLE_OS_MENU or not.

Hence, that's what you see in the code you commented on above.

Now I'm gathering I could just do `SoundLibrary::init()` in my prop somehow instead of all that runaround?
What would be the syntax for that, because I tried using the same line from prop_base `MKSPEC<MENU_SPEC_TEMPLATE>::SoundLibrary::init();` 
which says not in scope.
Then I tried to preface it with just `#define MENU_SPEC_TEMPLATE`
but then it says `wrong number of template arguments (0, should be 1)`

Reply:
> So the workaround was to first do #define MENU_SPEC_TEMPLATE DefaultMenuSpec at the top of my prop to  satisfy the prop_base stuff, then undef it, then re-define it according to BC_ENABLE_OS_MENU or not.

This doesn't work at all.
enterMenu() will do the wrong thing, and init() will be called on the wrong spec.

I think the way to solve this is to not set MENU_SPEC_TEMPLATE at all if your menus are being used.
You'll need to add an init call, like:

#ifndef MENU_SPEC_TEMPLATE
   MKSPEC<BcMenuSpec>::SoundLibrary::init();
#endif

This assumes that your menus are mutually exclusive with the prop_base ones, is that the case?

Basically, if the user defines MENU_SPEC_TEMPLATE it would disable your menues and use some other ones.

Setting MENUS_SPEC_TEMPLATE to BcMenuTemplate doesn't really work, since there is no TopMenu...

Reply:
ok. getting an error with that, error: invalid use of template-name 'MKSPEC' without an argument list
 1194 | MKSPEC::SoundLibrary::init();
 
 but once it's right, could it be like this so that BC_ENABLE_OS_MENU remains the thing to choose with?
```

#ifndef PROPS_SABER_BC_BUTTONS_H
#define PROPS_SABER_BC_BUTTONS_H

// #define MENU_SPEC_TEMPLATE DefaultMenuSpec
#ifdef BC_ENABLE_OS_MENU
#define MENU_SPEC_TEMPLATE DefaultMenuSpec
#else
MKSPEC::SoundLibrary::init();
#endif

#include "prop_base.h"
```

Reply:
Grr, it thinks I was writing HTML and hid it, let's see if this works better:

```cpp
MKSPEC<BcMenuSpec>::SoundLibrary::init();
```
The init() call needs to go in the prop constructor, or Setup() though, since it's a regular function call that needs to be executed.

> could it be like this so that BC_ENABLE_OS_MENU remains the thing to choose with?

I'm not sure I understand why you would want that?

> #define MENU_SPEC_TEMPLATE DefaultMenuSpec

Doing this makes it impossible for the user to select a different menu spec.
Currently there is only one, but that is likely to be short-lived.
Also, I want people to be able to put modified SPECs in their config files if they want to.

> Will the BCMenuSpec be known/used then without setting MENU_SPEC_TEMPLATE to it?

If your prop calls it, then it will be known/used, right?

> Still fuzzy on all this.

Well, the way I envisioned it is that prop makers don't really have to worry too much about the menus. My thinking was that there is one top-level menu, and all the prop has to do is to provide a key combination for entering the menu, the rest is up to the user, as they can then select which menu system and which sub-menus should be available by selecting a spec and which top menu to use.

However, I also want this to be as customizable as humanly possible, and my idea isn't the only way to do things. So if prop makers like you have ideas for how the menus should work, I want the system to be able to support that.  Since it's early days and nothing is set in stone yet, I'm more than happy to rewrite some stuff to make things work well for you.

The key piece that I feel that we're struggling with is separating "it's hard to make it work" from "i'm trying to do X". I'm more than happy to help with both, but it's much easier to do it well if I have a good understanding of what you are trying to achieve.

Reply:
> prop makers don't really have to worry too much about the menus. My thinking was that there is one top-level menu, and all the prop has to do is to provide a key combination for entering the menu, the rest is up to the user, as they can then select which menu system and which sub-menus should be available by selecting a spec and which top menu to use.

I guess I was trying to do the opposite. Make it a one define choice for the user to avoid swimming in a sea of too many options. Choose A and use the BC defaults, or B and use the full OS menu from top level down.
I don't want to restrict choice B though. I think it should be open for customization in the user config, just defaulted to the DefaultMenuSpec TopMenu.

Assuming there's going to be an easy-to-follow cheat sheet of sorts, as I'm not sure how many average users will be customising their menu experience on their own, and if they do want to, they'd want it as simple as possible. I think "MENU_SPEC_TEMPLATE" alone might be enough to scare off some people who would just rather choose A or B and be on their way. 
I, for one, can obviously appreciate all the customization options as a user and don't mind tailoring which menus to use and such. The flexibility is an awesome thing. But not everybody is like that.
So I'd like to find a middle ground, where using the BC prop gives you a default of custom menus (made for me because it's mine and that's how I'd like it), and if you want to customize, go for it!

Reply:
All that said, I think I got it. I updated the code, killed BC_ENABLE_OS_MENU, and updated the in-file documentation. 
I also did not know Setup() was a thing so that's great.